File: plot.py
import os
import numpy as np
import matplotlib.pyplot as plt


def plot_losses_per_experiment(
    loss_data, 
    output_dir, 
    log_scale=False, 
    log_y_scale=False, 
    extend_converged=True, 
    offset=0.2
):
    """
    Plot and save loss data for multiple experiments.
    
    Parameters:
        loss_data (list[dict]): A list of dictionaries where each dictionary maps algorithm names to loss values.
        output_dir (str): Directory to save the plots.
        log_scale (bool): Whether to use a log scale for the x-axis.
        log_y_scale (bool): Whether to use a log scale for the y-axis.
        extend_converged (bool): Whether to extend the curve horizontally after convergence.
        offset (float): Offset for vertical lines when algorithms converge at the same step.
    """
    os.makedirs(output_dir, exist_ok=True)
    colors = ['blue', 'green', 'red', 'orange', 'purple', 'cyan', 'magenta', 'black']
    line_offsets = offset  # Offset for vertical lines

    for i, experiment in enumerate(loss_data):
        plt.figure(figsize=(10, 6))
        max_iters = max(len(vals) for vals in experiment.values())

        # Collect convergence points
        convergence_points = {}

        for j, (algorithm, losses) in enumerate(experiment.items()):
            color = colors[j % len(colors)]
            iters = np.arange(1, len(losses) + 1)

            # Plot the losses for the algorithm
            plt.plot(iters, losses, label=algorithm, color=color)

            # Add a vertical line at the convergence point
            convergence_iter = len(losses)
            if convergence_iter in convergence_points:
                convergence_points[convergence_iter] += 1
            else:
                convergence_points[convergence_iter] = 1

            # Offset for vertical line to avoid overlap
            vertical_offset = line_offsets * convergence_points[convergence_iter]
            plt.axvline(x=convergence_iter + vertical_offset, linestyle='--', color=color)

            # Optionally extend the curve horizontally after convergence
            if extend_converged and len(losses) < max_iters:
                plt.plot(
                    np.arange(len(losses) + 1, max_iters + 1),
                    [losses[-1]] * (max_iters - len(losses)),
                    color=color,
                    linestyle='-',
                )

        # Customize the plot
        plt.xlabel("Iterations (log scale)" if log_scale else "Iterations")
        plt.ylabel("Objective Value (log scale)" if log_y_scale else "Objective Value")
        plt.title(f"Experiment {i + 1} Losses")
        if log_scale:
            plt.xscale('log')
        if log_y_scale:
            plt.yscale('log')
        plt.legend()

        # Save the plot
        output_path = os.path.join(output_dir, f"experiment_{i + 1}_losses.png")
        plt.savefig(output_path)
        plt.close()

File: plot.jl
using Plots

function plot_losses_per_experiment(
    loss_data::Vector{Dict{String, Vector{Float64}}}, 
    output_dir::String, 
    log_scale::Bool=false, 
    log_y_scale::Bool=false, 
    extend_converged::Bool=true,
    offset::Float64=0.2
)
    mkpath(output_dir)

    colors = [:blue, :green, :red, :orange, :purple, :cyan, :magenta, :black]
    line_offsets = offset  # Offset for vertical lines when algorithms converge at the same step

    for (i, experiment) in enumerate(loss_data)
        plot()  # Start a new plot for each experiment
        max_iters = maximum([length(vals) for vals in values(experiment)])
        
        # Collect convergence points
        convergence_points = Dict{Int, Int}()
        for (j, (algorithm, losses)) in enumerate(experiment)
            color = colors[j]
            iters = 1:length(losses)

            # Plot the losses for the algorithm
            plot!(iters, losses, label=algorithm, color=color)

            # Add a vertical line at the convergence point
            convergence_iter = length(losses)
            if haskey(convergence_points, convergence_iter)
                convergence_points[convergence_iter] += 1  # Increment count for this iteration
            else
                convergence_points[convergence_iter] = 1  # Initialize count
            end

            offset = line_offsets * convergence_points[convergence_iter]  # Offset for this iteration
            vline!([convergence_iter + offset], label="", linestyle=:dash, color=color)

            # Optionally extend the curve horizontally after convergence
            if extend_converged && length(losses) < max_iters
                plot!(length(losses):max_iters, fill(losses[end], max_iters - length(losses) + 1),
                      label="", color=color, linestyle=:solid)
            end
        end

        # Customize the plot
        xlabel!(log_scale ? "Iterations (log scale)" : "Iterations")
        ylabel!(log_y_scale ? "Objective Value (log scale)" : "Objective Value")
        title!("Experiment $i Losses")
        if log_scale
            xaxis!(:log10)
        end
        if log_y_scale
            yaxis!(:log10)
        end

        # Save the plot
        savefig(joinpath(output_dir, "experiment_$(i)_losses.png"))
    end
end
